frontend_vm/src/compiler.rs

```diff
@@ -0,0 +1,958 @@
+
+// Removes up to 16 bits beyond 32
+// TODO is this really safe?
+instr wrap16 Y -> X { Y = Y_b5 * 2**32 + Y_b6 * 2**40 + X, X = X_b1 + X_b2 * 0x100 + X_b3 * 0x10000 + X_b4 * 0x1000000 }
+pil{
+    col witness Y_b5;
+    col witness Y_b6;
+    { Y_b5 } in { bytes };
+    { Y_b6 } in { bytes };
+}
+
+// Removes up to 32 bits beyond 32
+// TODO is this really safe?
+instr mul Y, Z -> X {
+    Y * Z = X + Y_b5 * 2**32 + Y_b6 * 2**40 + Y_b7 * 2**48 + Y_b8 * 2**56,
+    X = X_b1 + X_b2 * 0x100 + X_b3 * 0x10000 + X_b4 * 0x1000000
+}
+// implements (Y * Z) >> 32
+instr mulhu Y, Z -> X {
+    Y * Z = X * 2**32 + Y_b5 + Y_b6 * 0x100 + Y_b7 * 0x10000 + Y_b8 * 0x1000000,
+    X = X_b1 + X_b2 * 0x100 + X_b3 * 0x10000 + X_b4 * 0x1000000
+}
+pil{
+    col witness Y_b7;
+    col witness Y_b8;
+    { Y_b7 } in { bytes };
+    { Y_b8 } in { bytes };
+}
+    "#
+}
+
+pub fn process_statement(s: Statement) -> Vec<String> {
+    match &s {
+        Statement::Label(l) => vec![format!("{}::", escape_label(l))],
+        Statement::Directive(directive, args) => match (directive.as_str(), &args[..]) {
+            (
+                ".loc",
+                [Argument::Expression(Expression::Number(file)), Argument::Expression(Expression::Number(line)), Argument::Expression(Expression::Number(column)), ..],
+            ) => {
+                vec![format!("  debug loc {file} {line} {column};")]
+            }
+            (".file", _) => {
+                // We ignore ".file" directives because they have been extracted to the top.
+                vec![]
+            }
+            _ if directive.starts_with(".cfi_") => vec![],
+            _ => panic!(
+                "Leftover directive in code: {directive} {}",
+                args.iter().map(|s| s.to_string()).join(", ")
+            ),
+        },
+        Statement::Instruction(instr, args) => process_instruction(instr, args)
+            .into_iter()
+            .map(|s| "  ".to_string() + &s)
+            .collect(),
+    }
+}
+
+pub fn quote(s: &str) -> String {
+    // TODO more things to quote
+    format!("\"{}\"", s.replace('\\', "\\\\").replace('\"', "\\\""))
+}
+
+pub fn escape_label(l: &str) -> String {
+    // TODO make this proper
+    l.replace('.', "_dot_").replace('/', "_slash_")
+}
+
+fn argument_to_number(x: &Argument) -> u32 {
+    if let Argument::Expression(expr) = x {
+        expression_to_number(expr)
+    } else {
+        panic!("Expected numeric expression, got {x}")
+    }
+}
+
+fn expression_to_number(expr: &Expression) -> u32 {
+    if let Expression::Number(n) = expr {
+        *n as u32
+    } else {
+        panic!("Constant expression could not be fully resolved to a number during preprocessing: {expr}");
+    }
+}
+
+fn argument_to_escaped_symbol(x: &Argument) -> String {
+    if let Argument::Expression(Expression::Symbol(symb)) = x {
+        escape_label(symb)
+    } else {
+        panic!("Expected a symbol, got {x}");
+    }
+}
+
+fn r(args: &[Argument]) -> Register {
```

Comment:
These functions here are probably generic enough, but we would need to export them, so they should probably go into a sub-module, I don't think it is a good idea to export single-letter functions :)